### PR TITLE
[RFC] shell.c: A full RBuffer with no NL can freeze shell output.

### LINKED ISSUE #2983: `:grep` with many files doesn't work
I tried to benchmark the `:grep` to compare it between vim and neovim.
But when I ran the commands on my [nvimfiles](https://github.com/kassio/nvimfiles) folder, with nvim, the process doesn't finish:

* **vim**
```console
$ vim --version
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Jul  8 2015 09:55:56)
MacOS X (unix) version
Included patches: 1-769
Compiled by Homebrew
#...

$ time vim +grep\ function +all
# ... shows found lines
vim +grep\ function +qall  0.63s user 0.23s system 83% cpu 1.030 total
```

* **nvim**
```console
$ nvim -version
NVIM 0.0.0-alpha+201507080742 (compiled Jul  8 2015 09:58:10)
Commit: ba2f7a9310bf3d3d0b028ad38e9724553e68c064
Build type: RelWithDebInfo
#...

$ time nvim +grep\ function +all
plugged/nerdtree/autoload/nerdtree/ui_glue.vim:249:3:" FUNCTION: s:findAndRevealPath() {{{1
plugged/nerdtree/autoload/nerdtree/ui_glue.vim:250:1:function! s:findAndRevealPath()
plugged/nerdtree/autoload/nerdtree/ui_glue.vim:297:4:endfunction
plugged/nerdtree/autoload/nerdtree/ui_glue.vim:299:2:"FUNCTION: s:handleLeftClick() {{{1
plugged/nerdtree/autoload/nerdtree/ui_glue.vim:301:1:function! s:handleLeftClick()
plugged/nerdtree/autoload/nerdtree/ui_glue.vim:305:63:        "the dir arrows are multibyte chars, and vim's string functions only
#... shows some more found lines, and then stop, freezes, doesn't finish the process.
```

### PR DESCRIPTION
**Pull request by oni-link**
shell.c: A full RBuffer with no NL can freeze shell output.

out_data_cb() can return without emptying the full RBuffer (no NL was
seen). Because the shell output stream is stopped until space in the
Rbuffer is freed up, no more shell output is written.

To prevent this, output the full RBuffer when write_output() did not
write anything.

write_output() can also process the same RBuffer content more than once,
if no NL was seen. To prevent NUL bytes from producing new lines (if
lines are not written to a buffer), translate NUL to SOH(1).

Fixes #2983

### CODE REVIEW

**Comment by justinmk** [comment]
Could these be related?

- https://github.com/neovim/neovim/issues/1714
- https://github.com/neovim/neovim/issues/1813

File: src/nvim/os/shell.c [hunk 1]
```diff
@@ -297,14 +297,22 @@ static void system_data_cb(Stream *stream, RBuffer *buf, void *data, bool eof)
...
-    if (written) {
-      rbuffer_consumed(buf, written);
-    } else {
-      break;
-    }
+  size_t cnt;
+  char *ptr = rbuffer_read_ptr(buf, &cnt);
+
+  if (!cnt) {
+      return;
```
Comment (tarruda) [inline src/nvim/os/shell.c]: This line has a 4 space indent

**Comment by tarruda** [comment]
Other than the lint issues, LGTM

> Could these be related?

No, those are separate issues

**Comment by justinmk** [comment]
Any ideas for a functional test?

**Comment by oni-link** [comment]
Repushed with fixed lint issues.

**Comment by justinmk** [comment]
For reference:

```
$ time vim -u NONE -c 'set grepprg=grep\ -r' -c 'grep void **/*' -c qall
real    0m5.078s
user    0m1.267s
sys     0m0.548s

$ time ./build/bin/nvim -u NONE -c 'set grepprg=grep\ -r' -c 'grep void **/*' -c qall                                                                       

real    0m11.267s
user    0m5.113s
sys     0m1.644s
```

**Comment by justinmk** [comment]
Merged, thank you @oni-link .

**Pull request closed without merging**